apps/src/applab/dontMarshalApi.js

```diff
@@ -96,83 +155,137 @@ export function removeItem(array, index) {
-  }
-}
+// ImageData RGB helper functions
 
-export function getBlue(imageData, x, y) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    return imageData.properties.data.properties[pixelOffset + 2].toNumber();
-  }
-}
+// TODO: more parameter validation (data array type, length), error output
 
-export function getAlpha(imageData, x, y) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    return imageData.properties.data.properties[pixelOffset + 3].toNumber();
+function getImageDataValue(
+  calledWithinInterpreter,
+  colorOffset,
+  imageData,
+  x,
+  y
+) {
+  const imageDataProperties = calledWithinInterpreter
+    ? imageData.properties
+    : imageData;
+  if (imageDataProperties.data && imageDataProperties.width) {
+    const pixelOffset = y * imageDataProperties.width * 4 + x * 4;
+    const totalOffset = pixelOffset + colorOffset;
+    return calledWithinInterpreter
+      ? imageDataProperties.data.properties[totalOffset].toNumber()
+      : imageDataProperties.data[totalOffset];
   }
 }
 
-export function setRed(imageData, x, y, value) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    imageData.properties.data.properties[pixelOffset] = value;
-  }
-}
+const interpreterGetRed = getImageDataValue.bind(null, true, 0);
+const interpreterGetGreen = getImageDataValue.bind(null, true, 1);
+const interpreterGetBlue = getImageDataValue.bind(null, true, 2);
+const interpreterGetAlpha = getImageDataValue.bind(null, true, 3);
 
-export function setGreen(imageData, x, y, value) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    imageData.properties.data.properties[pixelOffset + 1] = value;
-  }
-}
+export const getRed = getImageDataValue.bind(null, false, 0);
+export const getGreen = getImageDataValue.bind(null, false, 1);
+export const getBlue = getImageDataValue.bind(null, false, 2);
+export const getAlpha = getImageDataValue.bind(null, false, 3);
 
-export function setBlue(imageData, x, y, value) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    imageData.properties.data.properties[pixelOffset + 2] = value;
+function setImageDataValue(
+  calledWithinInterpreter,
+  colorOffset,
+  imageData,
+  x,
+  y,
+  value
+) {
+  const imageDataProperties = calledWithinInterpreter
+    ? imageData.properties
+    : imageData;
+  if (imageDataProperties.data && imageDataProperties.width) {
+    const pixelOffset = y * imageDataProperties.width * 4 + x * 4;
+    const totalOffset = pixelOffset + colorOffset;
+    if (calledWithinInterpreter) {
+      imageDataProperties.data.properties[totalOffset] = value;
```

Comment:
This may be unavoidable, but want to call out that we're explicitly depending on JS-Interpreter internal state (how it stores underlying data).

Reply:
absolutely - anything operating at this level (effectively trying to augment interpreter functionality) is subject to this. It may be avoidable in the long term if we want to create more of a formal interface around how to reach in to interpreter specific data.